x/gamm/pool-models/balancer/amm.go

```diff
@@ -88,6 +111,27 @@ func (p Pool) CalcInAmtGivenOut(
 	return sdk.NewDecCoinFromDec(tokenInDenom, tokenAmountInBeforeFee), nil
 }
 
+// SwapInAmtGivenOut is a mutative method for CalcOutAmtGivenIn, which includes the actual swap
+func (p Pool) SwapInAmtGivenOut(
+	ctx sdk.Context, tokensOut sdk.Coins, tokenInDenom string, swapFee sdk.Dec) (
+	tokenIn sdk.Coin, err error,
+) {
+	tokenInDecCoin, err := p.CalcInAmtGivenOut(ctx, tokensOut, tokenInDenom, swapFee)
+	if err != nil {
+		return sdk.Coin{}, sdkerrors.Wrapf(types.ErrInvalidMathApprox, "token amount is zero or negative")
+	}
+	tokenInCoin, _ := tokenInDecCoin.TruncateDecimal()
+	if tokenInCoin.Amount.LTE(sdk.ZeroInt()) {
+		return sdk.Coin{}, sdkerrors.Wrapf(types.ErrInvalidMathApprox, "token amount is zero or negative")
+	}
+
+	err = p.ApplySwap(ctx, sdk.Coins{tokenInCoin}, tokensOut)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+	return tokenInCoin, nil
+}
+
 // ApplySwap.
 func (p *Pool) ApplySwap(ctx sdk.Context, tokensIn sdk.Coins, tokensOut sdk.Coins) error {
```

Comment:
Lets make this unexposed 
```suggestion
func (p *Pool) applySwap(ctx sdk.Context, tokensIn sdk.Coins, tokensOut sdk.Coins) error {
```